xbmc/video/VideoInfoScanner.cpp

```diff
@@ -446,6 +452,120 @@ namespace KODI::VIDEO
+          if (UpdateSetInTag(movieDetails)) // Returned in tag
+          {
+            m_database.AddSet(movieDetails.m_set.GetTitle(), movieDetails.m_set.GetOverview(),
+                              movieDetails.m_set.GetOriginalTitle(),
+                              movieDetails.GetUpdateSetOverview()); // Update set
+            CGUIListItem::ArtMap movieSetArt;
+            auto urls = movieDetails.m_strPictureURL.GetUrls();
+            for (auto& url : urls)
+              if (StringUtils::StartsWith(url.m_aspect, "set."))
+                movieSetArt.insert({url.m_aspect.substr(4), url.m_url});
+            m_database.SetArtForItem(dbId, MediaTypeVideoCollection, movieSetArt);
+          }
+        }
+      }
+    }
+  }
+
+  bool CVideoInfoScanner::UpdateSetInTag(CVideoInfoTag& tag)
+  {
+    // Uses the set information in m_set from a tag of a movie
+    // If there is a set, see if the details need to be updated from the
+    // Movie Set Information Folder
+    bool setUpdated{false};
+    if (tag.m_set.HasTitle())
+    {
+      const std::string movieSetInfoPath = GetMovieSetInfoFolder(tag.m_set.GetTitle());
+      if (!movieSetInfoPath.empty())
+      {
+        // look for Set.NFO
+        if (const std::unique_ptr setLoader{
+                CSetInfoTagLoaderFactory::CreateLoader(tag.m_set.GetTitle())};
+            setLoader)
+        {
+          CSetInfoTag setTag;
+          INFO_TYPE result{setLoader->Load(setTag, false)};
+          if (result && !setTag.IsEmpty())
+          {
+            tag.m_set.SetOriginalTitle(tag.m_set.GetTitle());
+            if (!setTag.GetTitle().empty())
+              tag.m_set.SetTitle(setTag.GetTitle());
+            if (!setTag.GetOverview().empty())
+            {
+              tag.m_set.SetOverview(setTag.GetOverview());
+              tag.SetUpdateSetOverview(true);
+            }
+            if (!setTag.GetPoster().empty())
+              tag.m_set.SetPoster(setTag.GetPoster());
+            setUpdated = true;
+          }
+        }
+
+        // Now look for art
+        // If poster specified in set.nfo use that first
+        CGUIListItem::ArtMap movieSetArt;
+        if (tag.m_set.HasPoster())
+        {
+          movieSetArt.insert({"poster", tag.m_set.GetPoster()});
+        }
+        else
+        {
+          const std::vector<std::string> movieSetArtTypes =
+              CVideoThumbLoader::GetArtTypes(MediaTypeVideoCollection);
+          AddLocalItemArtwork(movieSetArt, movieSetArtTypes, movieSetInfoPath, true, false);
+        }
+        if (!movieSetArt.empty())
+        {
+          tag.m_strPictureURL.Parse();
+          auto urls = tag.m_strPictureURL.GetUrls();
+
+          // Clear existing art (if any) from cache and VideoInfoTag
+          CTextureDatabase textureDb;
+          if (textureDb.Open())
+          {
+            for (const auto& url : urls)
+              if (StringUtils::StartsWith(url.m_aspect, "set."))
```

Comment:
It may be I misunderstand exactly how the various caches work in Kodi. I thought as we are adding new artwork and not keeping the existing artwork it would be good practice to remove the existing artwork in the cache - is this not correct??

Reply:
Kodi has this cache so it doesn't download images from Fanart.TV and other webservices multiple times. This is throwing away cached images that Kodi may use, and will then have to download again.

Reply:
So you would expect the images to stay in the cache even if they are not currently referenced elsewhere, in case they were to be used again in the future??

Reply:
The images this currently invalidates should not be removed from the database - this is "all available artwork returned from the scraper".

Additionally, this code may even remove images that are in the new list, or images that might be referenced by something else, such as a movie or an alternate movie set name. All of these images could still be used now. Kodi 22 manages unreferenced images over time.